modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/js/index.js

```diff
@@ -6,64 +6,98 @@
 import {openModal} from 'frontend-js-components-web';
 import {addParams, fetch, navigate} from 'frontend-js-web';
 
-async function fetchModalContent(url) {
-	try {
-		const modalSignInURL = addParams('windowState=exclusive', url);
-
-		const response = await fetch(modalSignInURL);
-		const responseText = await response.text();
-
-		return responseText;
-	}
-	catch (error) {
-		if (process.env.NODE_ENV === 'development') {
-			console.error(error);
-		}
-
-		return '';
-	}
-}
-
 export function signInButtonPropsTransformer({
-	additionalProps: {redirect, signInURL},
+	additionalProps: {redirect: initialRedirect, signInURL},
 	...props
 }) {
-	const onClick = async () => {
-		if (redirect) {
-			navigate(signInURL);
+	const signInButton = document.querySelector('.sign-in > div > button');
+	const modalSignInURL = addParams('windowState=exclusive', signInURL);
+	let loading = false;
+	let modalContentHTML = '';
+	let isModalOpen = false;
+	let shouldRedirect = initialRedirect;
+
+	const updateModalContent = (html) => {
+		const modalBody = document.querySelector('.liferay-modal-body');
+		if (modalBody) {
+			const fragment = document
+				.createRange()
+				.createContextualFragment(html);
+			modalBody.innerHTML = '';
+			modalBody.appendChild(fragment);
+		}
+	};
 
+	const loadModalContent = () => {
+		if (loading || modalContentHTML) {
 			return;
 		}
 
-		const modalContentPromise = fetchModalContent(signInURL);
+		loading = true;
+
+		fetch(modalSignInURL)
+			.then((response) => response.text())
+			.then((responseHTML) => {
+				if (!loading) {
+					return;
+				}
 
-		openModal({
-			bodyHTML: '<span class="loading-animation"></span>',
-			containerProps: {className: ''},
-			onOpen: async () => {
-				const modalBody = document.querySelector('.liferay-modal-body');
+				loading = false;
 
-				try {
-					const modalContent = await modalContentPromise;
+				if (!responseHTML) {
+					shouldRedirect = true;
 
-					if (modalBody && modalContent) {
-						modalBody.innerHTML = modalContent;
-					}
-					else {
-						navigate(signInURL);
-					}
+					return;
 				}
-				catch (error) {
-					navigate(signInURL);
+
+				modalContentHTML = responseHTML;
+
+				if (isModalOpen) {
+					updateModalContent(responseHTML);
 				}
-			},
-			size: 'md',
-			title: Liferay.Language.get('sign-in'),
-		});
+			})
+			.catch(() => {
+				shouldRedirect = true;
+			});
 	};
 
 	return {
 		...props,
-		onClick,
+		onClick() {
+			loadModalContent();
+
+			if (shouldRedirect) {
+				navigate(signInURL);
+
+				return;
+			}
+
+			if (signInButton) {
+				openModal({
+					bodyHTML:
+						modalContentHTML || '<span class="loading-animation">',
```

Comment:
[NB] The browser (i.e. `innerHTML`) will automatically close this tag, even though this is malformed HTML (the tag should be explicitly closed).